java/src/main/java/ai/rapids/cudf/ColumnVector.java

```diff
@@ -331,234 +257,62 @@ public boolean hasNulls() {
-   * still return true.
-   */
-  public boolean hasHostData() {
-    return offHeap.getHostData() != null || rows == 0;
-  }
-
-  private void checkHasDeviceData() {
-    if (!hasDeviceData()) {
-      if (refCount <= 0) {
-        throw new IllegalStateException("Vector was already closed.");
-      }
-      throw new IllegalStateException("Vector not on device");
-    }
-  }
-
-  private void checkHasHostData() {
-    if (!hasHostData()) {
-      if (refCount <= 0) {
-        throw new IllegalStateException("Vector was already closed.");
-      }
-      throw new IllegalStateException("Vector not on host");
-    }
-  }
-
-  /**
-   * Drop any data stored on the host, but move it to the device first if necessary.
-   */
-  public final void dropHostData() {
-    ensureOnDevice();
-    if (offHeap.hostData != null) {
-      offHeap.hostData.close();
-      offHeap.hostData = null;
-      // Host data tracking happens on a per buffer basis.
-    }
-  }
-
-  /**
-   * Drop any data stored on the device, but move it to the host first if necessary.
+   * Copy the data to the host.
    */
-  public final void dropDeviceData() {
-    ensureOnHost();
-    if (offHeap.cudfColumnHandle != null) {
-      long amount = getDeviceMemorySize();
-      offHeap.cudfColumnHandle.close();
-      offHeap.cudfColumnHandle = null;
-      MemoryListener.deviceDeallocation(amount, internalId);
-    }
-  }
-
-  /**
-   * Be sure the data is on the device.
-   */
-  public final void ensureOnDevice() {
-    if (offHeap.cudfColumnHandle == null) {
-      if (rows == 0) {
-        offHeap.cudfColumnHandle = new CudfColumn(type);
-        MemoryListener.deviceAllocation(getDeviceMemorySize(), internalId);
-        return;
-      }
-      checkHasHostData();
-
-      assert type != DType.STRING || offHeap.getHostData().offsets != null;
-
-      try (DevicePrediction prediction =
-               new DevicePrediction(getHostMemorySize(), "ensureOnDevice");
-           NvtxRange toDev = new NvtxRange("ensureOnDevice", NvtxColor.BLUE)) {
-
-        offHeap.cudfColumnHandle = new CudfColumn(type, (int) rows, nullCount, offHeap.hostData.data,
-            offHeap.hostData.valid, offHeap.hostData.offsets);
-
-        if (offHeap.isLeakExpected()) {
-          offHeap.cudfColumnHandle.noWarnLeakExpected();
+  public HostColumnVector copyToHost() {
+    try (HostPrediction prediction =
+             new HostPrediction(getDeviceMemorySize(), "ensureOnHost");
+         NvtxRange toHost = new NvtxRange("ensureOnHost", NvtxColor.BLUE)) {
+      HostMemoryBuffer hostDataBuffer = null;
+      HostMemoryBuffer hostValidityBuffer = null;
+      HostMemoryBuffer hostOffsetsBuffer = null;
+      BaseDeviceMemoryBuffer valid = offHeap.getValid();
+      BaseDeviceMemoryBuffer offsets = offHeap.getOffsets();
+      BaseDeviceMemoryBuffer data = offHeap.getData();
+      boolean needsCleanup = true;
+      try {
+        if (!nullCount.isPresent()) {
```

Comment:
Nit: `getNullCount` already makes this check, so simpler to unconditionally call it.